# Functional Annotation Tools 


## Intuition
We will be dividing functional annotation tools into clustering, homology-based and ab-initio-based tools.

## Pipeline
We will be analyzing our DNA and protein sequences (in faa, fna, and gff files) using homology and ab-initio based techniques. We will be narrowing down the following categories to one tool based on efficiency and performance.e
### Clustering
- UCLUST
- CD-HIT
- LinClust
- MeShClust
### Homology
- Antibiotic Resistance
  - CARD
  - ResFinder
- Prophage
  - ProphET
  - PHASTER
- Virulence
  - VFDB
  - T3DB
- Operons
  - OperonDB
  - ODB
- Fully Automated Functional Annotation
  - PANZZER2
  - BLAST
- Plasmids
  - PlasmidSeeker
### Ab-initio
- Plasmids
  - PlasFlow
- Outer Membrane Protein
  - TBB-Prod
- Trans Membrane Protein
  - HMMTOP2
  - TMHMM2
- Signal Peptide
  - SignalP
  - tatP
  - Phobius
  - LipoP
- CRSIPR
  - PilerCR
  - CRT

## Goals
1. We need to be able to get good predictions for Functional Annotation to pass on to the Comparative Genomics group.
2. We need to quantify the performance of each tool based on more than just a simple histogram.
3. We need to finish the presentation before Spring break.
